AEXS/aex-141.md

```diff
@@ -300,124 +314,111 @@ ApprovalForAll(address, address, string)
 | :--- | :--- |
+| from | option(address) |
 | token_id | int |
+| data | option(string) |
 
-## Extension Swappable ("swappable")
+# Extensions
 
-### swap\(\)
+This section covers the extendability of the basic token - e.g. mintable, burnable.
 
-Burns the whole balance of `Call.caller` and stores the same amount in the `swapped` map. 
+When an NFT contract implements an extension its name should be included in the `aex141_extensions` array, in order for third party software or contracts to know the interface.
+Any extensions should be implementable without permission. Developers of extensions MUST choose a name for `aex141_extensions` that is not yet used. Developers CAN make a pull request to the reference implementation for general purpose extensions and maintainers choose to eventually include them.
 
-```sophia
-stateful entrypoint swap() : unit
-```
+## Extension Mintable ("mintable")
 
-| parameter | type |
-| :--- | :--- |
+The `mintable` extension SHOULD be used for generic NFT minting without specific requirements in regards to minting.
 
-| return | type |
-| :--- | :--- |
-| () | unit |
+### mint\(\)
+
+Issues a new token to the provided address. If the `owner` is a contract, `IAEX141Receiver.on_aex141_received` will be called with `data` if provided.
 
-### check_swap\(\)
+Emits the `Mint` event.
 
-Returns the amount of NFTs that were burned through `swap` for the provided account. 
+Throws if the call to `IAEX141Receiver.on_aex141_received` implementation failed (safe transfer)
 
 ```sophia
-stateful entrypoint check_swap(account: address) : int
+stateful entrypoint mint(owner: address, metadata: option(metadata), data: option(string)) : int
 ```
 
 | parameter | type |
 | :--- | :--- |
-| account | address |
+| owner | address |
+| metadata | option(metadata) |
+| data  | option(string) |
 
 | return | type |
 | :--- | :--- |
-| int | int |
-
-### swapped\(\)
+| token_id | int |
 
-Returns all of the swapped NFTs that are stored in contract state. 
+## Extension Burnable ("burnable")
 
-```sophia
-stateful entrypoint swapped() : map(address, int)
-```
+The `burnable` extension SHOULD be used if NFTs within a contract are intended to be burnable.
 
-| return | type |
-| :--- | :--- |
-| swapped | map(address, int) |
+### burn\(\)
 
-## Events
+Burns the NFT with the provided `token_id`.
 
-**Swap** - MUST trigger when tokens are swapped using the `swap` function.
+Emits the `Burn` event.
 
-The swap event arguments should be as follows: `(account,  value)`
+Throws if `Call.caller` is NOT the current owner or NOT approved to transfer on behalf of the owner.
 
 ```sophia
-Swap(address, int)
+stateful entrypoint burn(token_id: int) : unit
 ```
 
 | parameter | type |
 | :--- | :--- |
-| account| address |
-| value | int |
+| token_id | int |
 
-# Receiver contract interface
+## Extension Events
 
-## NFTReceiver
+**Mint**
+
+This event MUST be triggered whenever a new token is minted with the `mintable` extension.
```

Comment:
maybe the Event should be part of the main standard, as without minting you can never have nfts

Reply:
we had it in the main standard. I don't mind where to place. it's exactly sth. we discussed a lot and I am not very happy having to discuss it again.

anyway, IF we put it into the main standard every mint of an NFT using `mintable_template` extension will trigger both:
- `Mint`
- `TemplateMint`

Reply:
@jyeshe can you comment on that? would like to get aligned here

Reply:
@thepiwo this is related to this comment:
- https://github.com/aeternity/AEXs/pull/148#discussion_r996699326

Reply:
> @jyeshe can you comment on that? would like to get aligned here

On the first comment about when the MUST would apply to `Mint` event I would rephrase to:
`This event MUST be triggered whenever a new token is minted during contract `init` or with the `mintable` extension.

Reply:
On the `Mint` event to be part of the standard, I think @thepiwo is refering to putting on the event list. For me all events could be declared there, including `Burn` and `TemplateMint`, it would be helpful to show in the code a list of all events that the standard contract might emit and one thing less to adapt. I personally prefer have the option to remove rather than to add. But if `Burn` is not included I would put at least `Mint` along with `TemplateMint` because a collection might be comprised only by semi-fungible tokens as well. 

For each kind of minting it would continue to have a different event:
- `Mint` for (unique) non-fungible tokens
- `TemplateMint` for semi-fungible tokens

Reply:
@jyeshe I think we clarified that already from our side and future coding examples will also better show how extensions differ from the basic interface. we were just not able to show that in the past in a good and re-usable way. as for the standard definition, if an event does not belong to the core interface I wouldn't place it there. why should we?

if my interpretation of @thepiwo's statement is correct he would assume `Mint` to be emitted for **ANY** NFT, no matter if it's an NFT of a template or not. if we need to do that (for whatever reason), then we need to put it into the basic interface.

Reply:
> On the first comment about when the MUST would apply to `Mint` event I would rephrase to:
> `This event MUST be triggered whenever a new token is minted during contract `init`or with the`mintable` extension.

this means that `Mint` would (again) have to be part of the core interface. and how to handle minting on `mintable_template` extension then? emit both?
- `Mint`
- `TemplateMint`

Reply:
Just one event will be fine, as long as they are part of the standard

Reply:
is it fine then like proposed right now? let's resolve then

Reply:
Also if you wanna make the mintable not an extension anymore and only put the Mint on the data type I am ok with it. It's possible to check if it is template collection by the aex141_extensions method.

Reply:
Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it's possible to check if it is template collection by the aex141_extensions method.

`mint` would always fail though when a collection has only template nfts.

Reply:
> Just one event will be fine, as long as they are part of the standard

Could you elaborate on this? Unify Mint and TemplateMint?

Reply:
> It's an issue of nomenclature because things that can be sold multiple times might not be considered non-fungible.

in our case **EVERY** NFT will be unique :) just wanted to point that clearly out again. we do not have any "semi-fungible" tokens. at least it's not present in any proposal yet and I absolutely want to avoid introducing it to be honest 😅 

for sure people can think it's the same token. but it will have a serial no. and you can attach individual mutable attributes to each NFT


> Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it would be possible to check if it is template collection by the aex141_extensions method.
> 
> `mint` would always fail though when a collection has only template nfts.

I would still favor to keep it separate as it's a cleaner separation.


> Could you elaborate on this? Unify Mint and TemplateMint?

my understanding is that it's **EITHER** / **OR** as it's currently in the example implementation:
- `mintable` -> `Mint`
- `mintable_templates` -> `TemplateMint`

let's get this confirmed again by @thepiwo 😄

Reply:
I'm not sure what I'm supposed to confirm, but we can do it the way @marc0olo proposes

Reply:
> > It's an issue of nomenclature because things that can be sold multiple times might not be considered non-fungible.
> 
> in our case **EVERY** NFT will be unique :) just wanted to point that clearly out again. we do not have any "semi-fungible" tokens. at least it's not present in any proposal yet and I absolutely want to avoid introducing it to be honest 😅 
> 
> for sure people can think it's the same token. but it will have a serial no. and you can attach individual mutable attributes to each NFT
> 
> 
> > Also if you wanna make the mintable not an extension anymore and or only put the Mint on the data type it would be possible to check if it is template collection by the aex141_extensions method.
> > 
> > `mint` would always fail though when a collection has only template nfts.
> 
> I would still favor to keep it separate as it's a cleaner separation.
> 
> 
> > Could you elaborate on this? Unify Mint and TemplateMint?
> 
> my understanding is that it's **EITHER** / **OR** as it's currently in the example implementation:
> - `mintable` -> `Mint`
> - `mintable_templates` -> `TemplateMint`
> 
> let's get this confirmed again by @thepiwo 😄 

Well the whole template discussion started and ended up on code with the idea of selling multiple *copies* of the same nft including the same metadata. Arguable if they are really unique only because have different ID. But for the sake of communication I can call them also Template NFTs, no worries.

Reply:
let's postpone the metadata topic. let me resolve that conversation. let's get this PR merged :)